File: src/amortize/calculator.py
"""Amortization calculation logic."""

from decimal import ROUND_HALF_UP, Decimal
from typing import TypedDict


class AmortizationScheduleEntry(TypedDict):
    month: int
    payment: Decimal
    principal: Decimal
    interest: Decimal
    balance: Decimal


class AmortizationResult(TypedDict):
    monthly_payment: Decimal
    total_interest: Decimal
    schedule: list[AmortizationScheduleEntry]


def calculate_monthly_payment(
    principal: Decimal, annual_rate_pct: Decimal, term_months: int
) -> Decimal:
    """Calculate monthly payment using standard loan amortization formula.

    Args:
        principal: Loan amount in dollars
        annual_rate_pct: Annual interest rate as percentage (e.g., 6.0 for 6%)
        term_months: Number of months for the loan

    Returns:
        Monthly payment amount (rounded to cents)

    Raises:
        ValueError: If inputs are invalid
    """
    if principal <= 0:
        raise ValueError("principal must be > 0")
    if annual_rate_pct < 0 or annual_rate_pct > 50:
        raise ValueError("annual_rate_pct must be between 0 and 50")
    if term_months < 1 or term_months > 480:
        raise ValueError("term_months must be between 1 and 480")

    if annual_rate_pct == 0:
        return (principal / Decimal(term_months)).quantize(Decimal("0.01"), rounding=ROUND_HALF_UP)

    monthly_rate = annual_rate_pct / Decimal(100) / Decimal(12)

    numerator = principal * monthly_rate * ((1 + monthly_rate) ** term_months)
    denominator = ((1 + monthly_rate) ** term_months) - 1

    payment = (numerator / denominator).quantize(Decimal("0.01"), rounding=ROUND_HALF_UP)
    return payment


def calculate_amortization(
    principal: Decimal, annual_rate_pct: Decimal, term_months: int
) -> AmortizationResult:
    """Calculate full amortization schedule.

    Args:
        principal: Loan amount in dollars
        annual_rate_pct: Annual interest rate as percentage
        term_months: Number of months for the loan

    Returns:
        Dictionary with monthly_payment, total_interest, and full schedule
    """
    monthly_payment = calculate_monthly_payment(principal, annual_rate_pct, term_months)
    monthly_rate = Decimal(0)
    if annual_rate_pct > 0:
        monthly_rate = annual_rate_pct / Decimal(100) / Decimal(12)

    schedule: list[AmortizationScheduleEntry] = []
    balance = principal
    total_interest = Decimal(0)

    for month in range(1, term_months + 1):
        interest_payment = (balance * monthly_rate).quantize(
            Decimal("0.01"), rounding=ROUND_HALF_UP
        )

        if month == term_months:
            principal_payment = balance
            payment = principal_payment + interest_payment
        else:
            principal_payment = (monthly_payment - interest_payment).quantize(
                Decimal("0.01"), rounding=ROUND_HALF_UP
            )
            payment = monthly_payment

        balance = (balance - principal_payment).quantize(Decimal("0.01"), rounding=ROUND_HALF_UP)
        total_interest = (total_interest + interest_payment).quantize(
            Decimal("0.01"), rounding=ROUND_HALF_UP
        )

        schedule.append(
            {
                "month": month,
                "payment": payment,
                "principal": principal_payment,
                "interest": interest_payment,
                "balance": max(Decimal(0), balance),
            }
        )

    return {
        "monthly_payment": monthly_payment,
        "total_interest": total_interest,
        "schedule": schedule,
    }

File: apps/amortize-calculator/lib/calculator.ts
import Decimal from 'decimal.js';

export interface AmortizationScheduleEntry {
  month: number;
  payment: string;
  principal: string;
  interest: string;
  balance: string;
}

export interface AmortizationResult {
  monthly_payment: string;
  total_interest: string;
  schedule: AmortizationScheduleEntry[];
}

export function calculateMonthlyPayment(
  principal: Decimal,
  annualRatePct: Decimal,
  termMonths: number
): Decimal {
  if (principal.lte(0)) throw new Error('principal must be > 0');
  if (annualRatePct.lt(0) || annualRatePct.gt(50)) {
    throw new Error('annual_rate_pct must be between 0 and 50');
  }
  if (termMonths < 1 || termMonths > 480) {
    throw new Error('term_months must be between 1 and 480');
  }

  if (annualRatePct.eq(0)) {
    return principal.div(termMonths).toDecimalPlaces(2, Decimal.ROUND_HALF_UP);
  }

  const monthlyRate = annualRatePct.div(100).div(12);
  const onePlusRate = new Decimal(1).plus(monthlyRate);
  const numerator = principal.times(monthlyRate).times(onePlusRate.pow(termMonths));
  const denominator = onePlusRate.pow(termMonths).minus(1);

  return numerator.div(denominator).toDecimalPlaces(2, Decimal.ROUND_HALF_UP);
}

export function calculateAmortization(
  principal: Decimal,
  annualRatePct: Decimal,
  termMonths: number
): AmortizationResult {
  const monthlyPayment = calculateMonthlyPayment(principal, annualRatePct, termMonths);
  const monthlyRate = annualRatePct.gt(0)
    ? annualRatePct.div(100).div(12)
    : new Decimal(0);

  const schedule: AmortizationScheduleEntry[] = [];
  let balance = new Decimal(principal);
  let totalInterest = new Decimal(0);

  for (let month = 1; month <= termMonths; month++) {
    const interestPayment = balance
      .times(monthlyRate)
      .toDecimalPlaces(2, Decimal.ROUND_HALF_UP);

    let principalPayment: Decimal;
    let payment: Decimal;

    if (month === termMonths) {
      principalPayment = balance;
      payment = principalPayment.plus(interestPayment);
    } else {
      principalPayment = monthlyPayment
        .minus(interestPayment)
        .toDecimalPlaces(2, Decimal.ROUND_HALF_UP);
      principalPayment = Decimal.min(principalPayment, balance);
      payment = principalPayment.plus(interestPayment);
    }

    balance = balance.minus(principalPayment).toDecimalPlaces(2, Decimal.ROUND_HALF_UP);
    totalInterest = totalInterest.plus(interestPayment).toDecimalPlaces(2, Decimal.ROUND_HALF_UP);

    schedule.push({
      month,
      payment: payment.toFixed(2),
      principal: principalPayment.toFixed(2),
      interest: interestPayment.toFixed(2),
      balance: Decimal.max(balance, 0).toFixed(2),
    });
  }

  return {
    monthly_payment: monthlyPayment.toFixed(2),
    total_interest: totalInterest.toFixed(2),
    schedule,
  };
}
